source/common/upstream/upstream_impl.h

```diff
@@ -319,26 +310,61 @@ class HostDescriptionImpl : virtual public HostDescription,
   absl::optional<MonotonicTime> last_hc_pass_time_;
 };
 
+/**
+ * Final implementation of most of Upstream::HostDescription, providing const
+ * of the address-related member variables.
+ *
+ * See also LogicalHostDescriptionImpl in
+ * source/extensions/clusters/common/logical_host.h for a variant that allows
+ * safe dynamic update to addresses.
+ */
+class HostDescriptionImpl : public HostDescriptionImplBase {
+public:
+  HostDescriptionImpl(
+      ClusterInfoConstSharedPtr cluster, const std::string& hostname,
+      Network::Address::InstanceConstSharedPtr dest_address, MetadataConstSharedPtr metadata,
+      const envoy::config::core::v3::Locality& locality,
+      const envoy::config::endpoint::v3::Endpoint::HealthCheckConfig& health_check_config,
+      uint32_t priority, TimeSource& time_source, const AddressVector& address_list = {});
+
+  // HostDescription
+  Network::Address::InstanceConstSharedPtr address() const override { return address_; }
+  Network::Address::InstanceConstSharedPtr healthCheckAddress() const override {
+    return health_check_address_;
+  }
+  SharedConstAddressVector addressList() const override {
+    return (address_list_ == nullptr) ? std::make_shared<AddressVector>() : address_list_;
```

Comment:
Perf suggestion: instead of calling `make_shared<>` here, create a staticexpr object of a SharedConstAddressVector type and return it.
This avoids allocating an empty vector (and shared-pointer) by multiple calls.

Reply:
Nice idea. But we can't use constexpr for shared_ptr: `'shared_ptr<const std::vector<std::shared_ptr<const Envoy::Network::Address::Instance>>>' is not literal because its destructor is not constexpr`

I don't think it would be a perf win here to do a lazy-init static because that requires a hidden cross-thread lock and creating an empty vector should be cheap (one alloc).

If we had a context in which to work with we could store it there. Maybe the ClusterInfo? But I don't want to make the PR bigger.

We could in a follow up have this return a std::optional or allow nullptr returns but aht would make the PR bigger as well.

Reply:
Can we return return address_list_ as it is? And in the caller code, if some places need to derefence it, then add the null check there, and return emptry vector {}?  
Just feel a little bit odd to create a shared_ptr() in this read accessor call.

Reply:
I think it's safer to put this here rather than at (potentially) call-sites. But maybe if there's not many we could make them all check. I'll look at this.

Reply:
I'm experimentally working on a commit to change the API to be called `addressListOrNull` and shift the burden to call-sites.

Reply:
Yanjun this was a good call; we can move the checking to call-sites, there really was only one. I made it safe (hopefully) by renaming the function to addressListOrNull so it should be obvious.